packages/flutter/lib/src/material/theme_data.dart

```diff
@@ -1019,6 +1043,107 @@ class ThemeData extends Diagnosticable {
     properties.add(DiagnosticsProperty<ColorScheme>('colorScheme', colorScheme, defaultValue: defaultData.colorScheme));
     properties.add(DiagnosticsProperty<DialogTheme>('dialogTheme', dialogTheme, defaultValue: defaultData.dialogTheme));
     properties.add(DiagnosticsProperty<Typography>('typography', typography, defaultValue: defaultData.typography));
+    properties.add(DiagnosticsProperty<CupertinoThemeData>('cupertinoOverrideTheme', cupertinoOverrideTheme, defaultValue: defaultData.cupertinoOverrideTheme));
+  }
+}
+
+/// A [CupertinoThemeData] that defers unspecified theme attributes to an
+/// upstream Material [ThemeData].
+///
+/// This type of [CupertinoThemeData] is used by the Material [Theme] to
+/// harmonize the [CupertinoTheme] with the material theme's colors and text
+/// styles.
+///
+/// For a [CupertinoThemeData] whose unspecified attributes default to iOS
+/// defaults rather than to an ancestor Material [ThemeData], insert another
+/// [CupertinoTheme] as the `child` of the Material [Theme].
+///
+/// See also:
+///
+///  * [CupertinoThemeData], whose null constructor parameters default to
+///    reasonable iOS styling defaults rather than harmonizing with a Material
+///    theme.
+///  * [Theme], widget which inserts a [CupertinoTheme] with this
+///    [MaterialBasedCupertinoThemeData].
+// This class subclasses CupertinoThemeData rather than composes one because it
+// _is_ a CupertinoThemeData with partially altered behavior. e.g. its textTheme
+// is from the superclass and based on the primaryColor but the primaryColor
+// comes from the Material theme unless overridden.
+@immutable
+class MaterialBasedCupertinoThemeData extends CupertinoThemeData {
```

Comment:
fyi: having a subclass here makes CupertinoThemeData less efficient to access (every access now has to have a potential dynamic dispatch). If we could just create a regular CupertinoThemeData directly with the fallback logic in the caller, rather than having a subclass, it might be better for performance.